### PR TITLE
Add Question Answering Node

### PR DESCRIPTION
**Pull request by derkmed**
This PR should  be stacked on https://github.com/PTG-Kitware/angel_system/pull/297.

### CODE REVIEW

File: ros/angel_system_nodes/angel_system_nodes/question_answerer.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            colored_apology = colored(
+                "I'm sorry. I don't know how to answer your statement.", "light_red"
+            )
+            colored_emotion = colored(user_emotion, "light_red")
+            return_msg = (
+                f"{colored_apology} I understand that you feel {colored_emotion}."
+            )
+        return return_msg
+
+    def listener_callback(self, msg):
```
Comment (joshanderson-kw) [inline ros/angel_system_nodes/angel_system_nodes/question_answerer.py:113]:
```suggestion
    def user_emotion_callback(self, msg):
```

File: ros/angel_system_nodes/angel_system_nodes/question_answerer.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        Constant loop to process received messages.
+        """
+        while True:
+            msg = self.message_queue.get()
+            emotion = msg.user_emotion
+            response = self.get_response(msg.utterance_text, emotion)
+            self.publish_generated_response(msg.utterance_text, response)
+
+    def publish_generated_response(self, utterance: str, response: str):
+        msg = SystemTextResponse()
```
Comment (joshanderson-kw) [inline ros/angel_system_nodes/angel_system_nodes/question_answerer.py:134]: Also fill in ROS header here (frame ID and timestamp)

File: ros/angel_system_nodes/angel_system_nodes/question_answerer.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        for p in set_parameters:
+            if p.type_ is rclpy.parameter.Parameter.Type.NOT_SET:
+                some_not_set = True
+                self.log.error(f"Parameter not set: {p.name}")
+        if some_not_set:
+            raise ValueError("Some parameters are not set.")
+
+        self._in_emotion_topic = self.get_parameter(IN_EMOTION_TOPIC).value
+        self._out_qa_topic = self.get_parameter(OUT_QA_TOPIC).value
+        self.prompt_file = self.get_parameter(FEW_SHOT_PROMPT).value
```
Comment (joshanderson-kw) [inline ros/angel_system_nodes/angel_system_nodes/question_answerer.py:41]: What is the purpose of this prompt file?

File: ros/angel_system_nodes/angel_system_nodes/question_answerer.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            f"({type(self._out_qa_topic).__name__}) "
+            f"{self._out_qa_topic}"
+        )
+        self.log.info(
+            f"Few-shot prompt file: "
+            f"({type(self.prompt_file).__name__}) "
+            f"{self.prompt_file}"
+        )
+
+        self.message_queue = queue.Queue()
```
[the remark this answers is not in the document]
Comment (derkmed) [reply ros/angel_system_nodes/angel_system_nodes/question_answerer.py:58]: changed to `question_queue`

File: ros/angel_system_nodes/angel_system_nodes/question_answerer.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+FEW_SHOT_PROMPT = "few_shot_prompt_file"
+
+
+class QuestionAnswerer(Node):
+    def __init__(self):
+        super().__init__(self.__class__.__name__)
+        self.log = self.get_logger()
+
+        parameter_names = [IN_EMOTION_TOPIC, OUT_QA_TOPIC, FEW_SHOT_PROMPT]
+        set_parameters = self.declare_parameters(
```
[the remark this answers is not in the document]
Comment (derkmed) [reply ros/angel_system_nodes/angel_system_nodes/question_answerer.py:27]: Done.

**Comment by derkmed** [comment]
This PR should be blocked on https://github.com/PTG-Kitware/angel_system/pull/312

File: ros/angel_system_nodes/angel_system_nodes/question_answerer.py (hunk 3 above)
Comment (derkmed) [reply ros/angel_system_nodes/angel_system_nodes/question_answerer.py]: I felt it was a cleaner organization of the prompt text. All LLM prompts get parsed into the format listed in the specified PROMPT file.

Alternatively we could have a very long python string, which doesn't appear as nice and adds more lines to the .py file.

File: ros/angel_system_nodes/resource/question_answer_site_demo_prompt
Comment (joshanderson-kw) [inline ros/angel_system_nodes/resource/question_answer_site_demo_prompt:1]: We may want to move this out of the `resource` directory and into `configs`. I think `resource` is a folder used by ROS when creating the `angel_system_nodes` package.

Comment (derkmed) [reply ros/angel_system_nodes/resource/question_answer_site_demo_prompt:1]: Done.

Comment (joshanderson-kw) [reply ros/angel_system_nodes/resource/question_answer_site_demo_prompt:1]: Hmm I am still seeing it under the `resource` directory.

**Review by joshanderson-kw — APPROVED** [review]

**Pull request merged**